design/design-doc-operations.md

```diff
@@ -0,0 +1,501 @@
+
+```proto
+message Requirements {
+  map<string, ResourceSelector> required_resources = 1;
+}
+
+message RunFunctionRequest {
+    // All other fields omitted.
+
+    map<string, Resources> required_resources = 6;
+}
+
+```
+
+Renaming a protobuf message field isn't a breaking change, so older function
+SDKs that use e.g. `GetExtraResources()` will continue to work.
+
+An operation function can instruct Crossplane to create or update[^1] arbitary
+resources by including server-side apply [fully-specified intent][8] (FSI)
+patches in `rsp.desired.resources`, just like a composition function.
+
+Each unique Operation will be the server-side apply field manager of any applied
+fields, and the Operation controller will force conflicts. This means the
+Operation controller will assume management of a field that's already set, and
+overwrite its value.
+
+This has a few implications:
+
+1. Operations must take care not to fight over fields with XRs and other
+   Operations. For example if an XR and an Operation both try to own an MR's
+   `spec.forProvider.version` field, they'll enter an endless loop fighting over
+   it.
+1. Unlike an XR, an Operation can't delete a field simply by omitting it from
+   its desired state (i.e. SSA FSI). XRs delete fields by first setting them in
+   FSI, then later omitting them from FSI. Operations will only run once.
+   They'll need to delete fields by explicitly setting them to `null`.
+
+An Operation's pipeline will run to completion once, as soon as you create it.
+I propose we support three ways - in addition to manual creation - to create an
+Operation:
+
+1. Compose an Operation using an XR
+1. Create an Operation on a regular schedule using a CronOperation
+1. Create an Operation when a resource changes using a WatchOperation
+
+Composing an Operation is hopefully self-explanatory. Operation doesn't have an
+abstraction resource (like an XR) by design - see [Alternatives
+Considered][#alternatives-considered] for more details.
+
+CronOperation and WatchOperation will be to Operation as CronJob is to Job in
+Kubernetes. They'll contain a template for an Operation, and create one as
+needed. Either on a cron schedule, or when a watched resource changes.
+
+Here's an example CronOperation:
+
+```yaml
+apiVersion: daytwo.crossplane.io/v1alpha1
+kind: CronOperation
+metadata:
+  name: cluster-rolling-upgrade
+spec:
+  # Crontab schedule
+  schedule: "0 12 * * *"
+  # How long after the scheduled time to wait before considering it too late to
+  # create the Operation.
+  startDeadline: 10m
+  # How many completed Operations to keep around.
+  successfulHistoryLimit: 3
+  failedHistoryLimit: 3
+  # Specifies how to treat concurrent executions of an operation - Allow
+  # (default), Forbid, or Replace.
+  concurrencyPolicy: Forbid
+  operationTemplate:
+    spec:
+      # How many times the pipeline can fail before the Operation fails.
+      failureLimit: 5
```

Comment:
`retryLimit`?

Reply:
Interestingly Job uses `backoffLimit`, which feels less intuitive to me than `retryLimit`. I'll use `retryLimit`